source/extensions/filters/http/fault/fault_filter.cc

```diff
@@ -169,6 +170,7 @@ bool FaultFilter::maybeSetupDelay(const Http::RequestHeaderMap& request_headers)
     delay_timer_->enableTimer(duration.value(), &decoder_callbacks_->scope());
     recordDelaysInjectedStats();
     decoder_callbacks_->streamInfo().setResponseFlag(StreamInfo::ResponseFlag::DelayInjected);
+    decoder_callbacks_->streamInfo().dynamicMetadata().MergeFrom(fault_settings_->metadata());
```

Comment:
Can we set metadata through `setDynamicMetadata` with namespace used in reverse DNS format

Reply:
How do we know the namespace? And do we know if the provided metadata has only one filter namespace ? It could have many filter related namespace + structs right?

Reply:
Here, it is `envoy.filters.http.fault`. `envoy.` namesapce is save for built-in filer. 
For non built-in filter, you can find example in google3's ilb/configuration_constants.h and how it is used to mutate dynamic metadata.

Metadata is a map of proto struct. The filter namespace is unique and fixed to each filter.  It is also why namespace (as key)  is needed: when multiple filters in filter chain, you need this key to find the associated metadata to use/mutate.

Reply:
Thats not what I am saying. If you look at the route metadata, it is plainly merged into dynamic metadata. It is not automagically suffixed under envoy.filters.http.router. The guidance is that users should specify the filter namespace in the metadata. And if you look at the tests, you will see that there is a filter namespace under which the protobuf struct is added. So I feel that there is no need for us to add an additional level into this dynamic metadata.

Reply:
dynamic metadata in streamInfo can be shared with different filters in the chain. Underlying it is a map `map<string, google.protobuf.Struct>`

Think about the case here: `filter_a->filter_b ->filter_c ((you are here)) ->filter_d` what you are doing here is override all other filters's metadata with your filter_c's metadata. Instead, you should only update the data associated with your filter.
route metadata is different I think it is not used to be shared between different filters.

Reply:
Note that we are merging and not overwriting. So we will not overwrite any other filter's dynamic metadata as long as the namespaces they use do not collide. If two filters have the same metadata namespace, then the contents will be merged using usual protobuf struct merge semantics. This is the same principle as Route metadata.

Reply:
You are right that as long as the namespaces do not collide, it won't not overwrite.

However, I think it is wrong for the case that name space collide. It will be overwrite if. The specific test `DynamicMetadataTest` in stream_info_impl_test.cc can be used to show this overwrite behavior.

So basically, you are override the filter state of fault_filter every time `MergeFrom` is invoked.

Reply:
Actually,  think we do need to fix this. We should merge dynamic metadata here under the configured filter name as key, not on top of all existing dynamic metadata. That's broken IMHO as there is too much chance of conflict and it doesn't properly namespace.